pinot-core/src/main/java/org/apache/pinot/core/plan/SelectionPlanNode.java

```diff
@@ -32,66 +30,65 @@
-    if (_selection.getSize() > 0) {
-      if (_selection.getSelectionSortSequence() == null) {
-        return new SelectionOnlyOperator(_indexSegment, _selection, transformOperator);
+    Selection selection = _queryContext.getBrokerRequest().getSelections();
+    if (_queryContext.getLimit() > 0) {
+      if (_queryContext.getOrderByExpressions() == null) {
+        return new SelectionOnlyOperator(_indexSegment, selection, transformOperator);
       } else {
-        return new SelectionOrderByOperator(_indexSegment, _selection, transformOperator);
+        return new SelectionOrderByOperator(_indexSegment, selection, transformOperator);
       }
     } else {
-      return new EmptySelectionOperator(_indexSegment, _selection, transformOperator);
+      return new EmptySelectionOperator(_indexSegment, selection, transformOperator);
     }
   }
 
-  private Set<TransformExpressionTree> collectExpressionsToTransform(IndexSegment indexSegment,
-      BrokerRequest brokerRequest) {
-
-    Set<TransformExpressionTree> expressionTrees = new LinkedHashSet<>();
-    Selection selection = brokerRequest.getSelections();
+  private Set<TransformExpressionTree> collectExpressions() {
+    Set<TransformExpressionTree> expressionTrees = new HashSet<>();
 
     // Extract selection expressions
-    List<String> selectionColumns = selection.getSelectionColumns();
-    if (selectionColumns.size() == 1 && selectionColumns.get(0).equals("*")) {
-      for (String column : indexSegment.getPhysicalColumnNames()) {
-        expressionTrees.add(new TransformExpressionTree(new IdentifierAstNode(column)));
+    List<ExpressionContext> selectExpressions = _queryContext.getSelectExpressions();
+    if (selectExpressions.size() == 1 && selectExpressions.get(0).equals(SelectionOperatorUtils.IDENTIFIER_STAR)) {
+      for (String column : _indexSegment.getPhysicalColumnNames()) {
+        expressionTrees
+            .add(new TransformExpressionTree(TransformExpressionTree.ExpressionType.IDENTIFIER, column, null));
       }
     } else {
-      for (String selectionColumn : selectionColumns) {
-        expressionTrees.add(TransformExpressionTree.compileToExpressionTree(selectionColumn));
+      for (ExpressionContext selectExpression : selectExpressions) {
+        expressionTrees.add(selectExpression.toTransformExpressionTree());
       }
     }
 
-    // Extract order-by expressions.
-    if (selection.getSize() > 0) {
-      List<SelectionSort> sortSequence = selection.getSelectionSortSequence();
-      if (sortSequence != null) {
-        for (SelectionSort selectionSort : sortSequence) {
-          String orderByColumn = selectionSort.getColumn();
-          expressionTrees.add(TransformExpressionTree.compileToExpressionTree(orderByColumn));
-        }
+    // Extract order-by expressions
+    List<OrderByExpressionContext> orderByExpressions = _queryContext.getOrderByExpressions();
+    if (_queryContext.getLimit() > 0 && orderByExpressions != null) {
```

Comment:
Why do we have to check for limit to extract order by expressions?

Reply:
LIMIT 0 is solved with EmptySelectionOperator where order-by expressions are not fetched. Added some notes here.